contracts/ProviderManager.sol

```diff
@@ -0,0 +1,41 @@
+pragma solidity ^0.4.24;
+
+contract ProviderManager {
+  event ProviderAdded (
+    address indexed provider,
+    uint pricePerStorageMineral,
+    uint pricePerComputeMineral,
+    uint blockRewardCut,
+    uint feeShare
+  );
+
+  event ProviderUpdated (
+    address indexed provider,
+    uint pricePerStorageMineral,
+    uint pricePerComputeMineral,
+    uint blockRewardCut,
+    uint feeShare
+  );
+
+  event ProviderResigned (
+    address indexed provider
+  );
+
+  enum ProviderStatus { Unregistered, Registered }
+
+  struct Provider {
+    uint pricePerStorageMineral;
+    uint pricePerComputeMineral;
+    uint blockRewardCut;
+    uint feeShare;
+  }
+
+  uint public numberOfProviders;
+  // TODO: rename to registeredProviders
```

Comment:
why not do this now?

Reply:
This PR already contains a good amount of refactoring, I'd rather spread it because it affects negatively the readability of the PR, and takes the attention off of the important parts